core/drivers/clk/clk.c

```diff
@@ -17,25 +17,27 @@ static unsigned int clk_lock = SPINLOCK_UNLOCK;
 struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
 		      struct clk **parent_clks, size_t parent_count)
 {
-	struct clk *clk = NULL;
+	struct clk_elt *clk_elt = NULL;
 	size_t parent = 0;
 
+	assert(ops->id == CLK_OPS_ELT);
 	if (name && !ops->get_name)
 		DMSG("Clock name unregistered for %s", name);
 
-	clk = calloc(1, sizeof(*clk) + parent_count * sizeof(clk));
-	if (!clk)
+	clk_elt = calloc(1, sizeof(*clk_elt) +
+			 parent_count * sizeof(*parent_clks));
+	if (!clk_elt)
 		return NULL;
 
-	clk->num_parents = parent_count;
+	clk_elt->num_parents = parent_count;
 	for (parent = 0; parent < parent_count; parent++)
-		clk->parents[parent] = parent_clks[parent];
+		clk_elt->parents[parent] = parent_clks[parent];
 
-	clk->name = name;
-	clk->ops = ops;
-	refcount_set(&clk->enabled_count, 0);
+	clk_elt->clk.name = name;
+	clk_elt->clk.ops = ops;
+	refcount_set(&clk_elt->clk.enabled_count, 0);
 
-	return clk;
+	return &clk_elt->clk;
 }
 
 void clk_free(struct clk *clk)
```

Comment:
What do you mean? Memory allocated from `clk_alloc()` are released by `clk_free()` without changes.
This comment is still valid with the raw/lw clock introduced later (allocated from `clk_raw/lw_alloc()`).
Did I miss something?.

Reply:
The `clk` is allocated in `clk_alloc()` as a `struct clk_elt`, but here you free it as a `struct clk` without first converting it to it's container struct. This happens to work here because `clk` is the first element in `struct clk_elt`, but that seems like an unexpected dependency.

Reply:
Actually, I expected the dependency and though it would be ok to use such shortcut.
I understand your concern. I'll update. Thanks for catching that.